ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/tilize_with_val_padding_op.cpp

```diff
@@ -17,21 +17,23 @@ void TilizeWithValPadding::validate(const std::vector<Tensor>& input_tensors) co
     TT_FATAL(input_tensor_a.get_layout() == Layout::ROW_MAJOR, "Can only tilize row major data");
     TT_FATAL(input_tensor_a.get_dtype() == DataType::BFLOAT16);
 
-    TT_FATAL(input_tensor_a.get_legacy_shape()[0] <= this->output_tensor_shape[0]);
-    TT_FATAL(input_tensor_a.get_legacy_shape()[1] <= this->output_tensor_shape[1]);
-    TT_FATAL(input_tensor_a.get_legacy_shape()[2] <= this->output_tensor_shape[2]);
-    TT_FATAL(input_tensor_a.get_legacy_shape()[3] <= this->output_tensor_shape[3]);
+    const auto& input_shape = input_tensor_a.get_legacy_shape();
+    for (auto i = 0; i < input_shape.rank(); i++) {
+        TT_FATAL(input_shape[i] <= this->output_tensor_shape[i],
+                 "Output shape {} must be greater than or equal to input shape {} in each dimension",
+                 this->output_tensor_shape, input_shape);
+    }
 
-    uint32_t num_rows = this->output_tensor_shape[2];
-    uint32_t inner_dim = this->output_tensor_shape[3];
-    TT_FATAL(num_rows % TILE_HEIGHT == 0, "Output shape must be tilizable");
-    TT_FATAL(inner_dim % TILE_WIDTH == 0, "Output shape must be tilizable");
+    uint32_t num_rows = this->output_tensor_shape[input_shape.rank() - 1];
+    uint32_t inner_dim = this->output_tensor_shape[input_shape.rank() - 2];
+    TT_FATAL(num_rows % TILE_HEIGHT == 0, "Output shape must be tilizable, but {} is not divisible by {}", num_rows, TILE_HEIGHT);
+    TT_FATAL(inner_dim % TILE_WIDTH == 0, "Output shape must be tilizable, but {} is not divisible by {}", inner_dim, TILE_WIDTH);
 
     if (input_tensor_a.memory_config().is_sharded()) {
         TT_FATAL(input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED);
         TT_FATAL(this->output_mem_config.memory_layout == input_tensor_a.memory_config().memory_layout);
-        for (uint32_t i = 0; i < input_tensor_a.get_legacy_shape().rank(); i++) {
-            if (i != input_tensor_a.get_legacy_shape().rank() - 2) {
+        for (uint32_t i = 0; i < input_shape.rank(); i++) {
+            if (i != input_shape.rank() - 2) {
```

Comment:
do we guarantee that rank() >= 2? 
nit: In general, i prefer to check i + 2 != input_shape.rank(). this way we will never go out of bounds for unsigned types.

Reply:
thanks!

Reply:
In general we can't tilize tensors with shape rank < 2 today. I am adding a check

Reply:
Also... why check the whole [0..rank] range if you can check [0..rank-2] range

Reply:
`for (auto i = 0; i + 2 < input_shape.rank(); ++i)` looks like what we need.

Reply:
`rank` has `size_t` type. At some point of time, we should discuss whether we want 64-bit type everywhere or uint32_t